delivery_roulier_picking_batch/models/stock_picking.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2024 Akretion (http://www.akretion.com).
+# @author Florian Mounier <florian.mounier@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from functools import reduce
+
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _is_batch_roulier(self):
+        # Check if the picking is part of a batch with a roulier carrier
+        self.ensure_one()
+        return (
+            self.batch_id
+            and self.batch_id.carrier_id
+            and self.batch_id.carrier_id._is_roulier()
+        )
+
+    def send_to_shipper(self):
+        self.ensure_one()
+        if self._is_batch_roulier():
+            # We are in a batch with a roulier carrier
+            # We need to send unsent packages
+            packages = self.package_ids - self.batch_id.sent_package_ids
+            if not packages:
+                # Nothing to send
+                return
+
+            # First sanity checks
+            # Check that all package pickings have the same sender/receiver:
+            for package in packages:
+                package_pickings = (
+                    self.env["stock.move.line"]
+                    .search(
+                        [
+                            "|",
+                            ("result_package_id", "=", package.id),
+                            ("package_id", "=", package.id),
+                        ]
+                    )
+                    .mapped("picking_id")
+                )
```

Comment:
We can make it simplier.
We should just check that all picking of the batch have the same delivery address

Reply:
I'd rather work at package level since it's them that will be sent and you may want to send 2 packages from different pickings to different addresses.

Reply:
In a functional point of view. When you use the batch delivery carrier you want to prepare all the picking for the same address, if user can mix, it will be a nightmare ;).
I am in favors of forcing to set a partner_id on the stock.picking.batch so it will be more understandable for the customer